flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.common.typeutils;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.IntStream;
+
+/**
+ * Base class for composite serializers.
+ *
+ * <p>This class serializes a composite type using array of its field serializers.
+ * Fields are indexed the same way as their serializers.
+ *
+ * @param <T> type of custom serialized value
+ */
+public abstract class CompositeSerializer<T> extends TypeSerializer<T> {
+	private static final long serialVersionUID = 1L;
+
+	/** Serializers for fields which constitute T. */
+	protected final TypeSerializer<Object>[] fieldSerializers;
+
+	/** Whether T is an immutable type. */
+	final boolean immutableTargetType;
+
+	/** Byte length of target object in serialized form. */
+	private final int length;
+
+	/** Whether any field serializer is stateful. */
+	private final boolean stateful;
+
+	private final int hashCode;
+
+	@SuppressWarnings("unchecked")
+	protected CompositeSerializer(boolean immutableTargetType, TypeSerializer<?> ... fieldSerializers) {
+		Preconditions.checkNotNull(fieldSerializers);
+		Preconditions.checkArgument(Arrays.stream(fieldSerializers).allMatch(Objects::nonNull));
+		this.immutableTargetType = immutableTargetType &&
+			Arrays.stream(fieldSerializers).allMatch(TypeSerializer::isImmutableType);
+		this.fieldSerializers = (TypeSerializer<Object>[]) fieldSerializers;
+		this.length = calcLength();
+		this.stateful = isStateful();
+		this.hashCode = Arrays.hashCode(fieldSerializers);
```

Comment:
I think up to this point, the code is iterating `fieldSerializers` 5 times (null checks, immutable check, length calc, stateful check, and hash code computation. It could be done in one iteration, but since this method should typically not be called in hot loops, this is an optional improvement.